Sagemaker orchestrator integration (#1177)
zenml-io/zenml: docs/book/component-gallery/orchestrators/sagemaker.md

---
description: How to orchestrate pipelines with Amazon Sagemaker
---

The Sagemaker orchestrator is an [orchestrator](./orchestrators.md) flavor provided

> with the ZenML `aws` integration that uses [Amazon Sagemaker](https://aws.amazon.com/sagemaker)
> to run your pipelines.

Reviewer: ```suggestion
with the ZenML `aws` integration that uses [Amazon Sagemaker Pipelines](https://sagemaker-examples.readthedocs.io/en/latest/sagemaker-pipelines/index.html)
to run your pipelines.
```

* you're looking for a UI in which you can track your pipeline runs.
* you're looking for a managed solution for running your pipelines.
* you're looking for a serverless solution for running your pipelines.

## How to deploy it

In order to use a Sagemaker AI orchestrator, you need to first deploy [ZenML to the
cloud](../../getting-started/deploying-zenml/deploying-zenml.md). It would be
recommended to deploy ZenML in the same region as you plan on using for
Sagemaker, but it is not necessary to do so. You must ensure that you are
[connected to the remote ZenML
server](../../starter-guide/collaborate/zenml-deployment.md) before using this
stack component.

The only other thing necessary to use the ZenML Sagemaker orchestrator is
enabling the relevant permissions for your particular role.

In order to quickly enable APIs, and create other resources necessary for to use
this integration, we will soon provide a Sagemaker stack
recipe via [our `mlops-stacks` recipe repository](https://github.com/zenml-io/mlops-stacks), which
will help you set up the infrastructure with one click.

## How to use it

To use the Sagemaker orchestrator, we need:

> * The ZenML `aws` integration installed. If you haven't done so, run 
>     ```shell
>     zenml integration install aws
>     ```

Reviewer: ```suggestion
* The ZenML `aws` integration installed. If you haven't done so, run 

```shell
zenml integration install aws
```
```

* [Docker](https://www.docker.com) installed and running.

> * [kubectl](https://kubernetes.io/docs/tasks/tools/#kubectl) installed.

Reviewer: Why do we need kubectl?

Author: Not sure. I started with the Vertex docs. Do we not need it? It would make sense that we don't need it, but I'm not sure if it's used somewhere that I didn't touch / know about in the codebase.

Reviewer: It shouldnt even be in the Vertex docs, so if its there maybe remove it there too :-)

* A [remote artifact store](../artifact-stores/artifact-stores.md) as part of 
your stack (configured with an `authentication_secret` attribute) along with [a remote secrets manager](../secrets-managers/secrets-managers.md) to store that
authentication secret.
* A [remote container registry](../container-registries/container-registries.md) 

> as part of your stack.

Reviewer: We also need the local client of the runner (the person who does python run.py) to have a role that can launch jobs in Sagemaker

> * An IAM role or user with an `AmazonSageMakerFullAccess` policy applied to it

Reviewer: Isnt this a bit too open? Perhaps we can find a better policy to suggest?

Author: This contains everything we need except for access to the secrets manager. We could probably find a more selective subset of the permissions contained in `AmazonSagemakerFullAccess` but it is the one Amazon recommends when people create roles created for using Sagemaker. I think it's not an unreasonable set of permissions to use.

  as well as `sagemaker.amazonaws.com` added as a Principal Service. Full
  details on these permissions can be found
  [here](https://docs.aws.amazon.com/sagemaker/latest/dg/sagemaker-roles.html)

>   or use the ZenML recipe (when available) which will set up the necessary

Reviewer: Correct me if im wrong, but doesnt the role also need access to the container registry, secrets manager, and the artifact store? How else would it be interacting with these things? Maybe mention here that you need to add the neccessary permissions to the sagemaker role

Author: All (apart from secrets manager access) are all included with the `AmazonSageMakerFullAccess`. I tested this in a fresh account yesterday.

>     permissions for you.

Reviewer: I think this part needs to be fleshed out. How do you actually enable these pipelines and see the UI? I remember we went through so many hoops to do it, and id imagine it would be nice to understand how to do it for our users too. Also I think you need to create a Sagemaker role for the pipeline as you do it

Author: https://docs.aws.amazon.com/sagemaker/latest/dg/security-iam-awsmanpol.html has more on their managed policies.

which includes your code and use it to run your pipeline steps in Sagemaker. 
Check out [this page](../../advanced-guide/pipelines/containerization.md)
if you want to learn more about how ZenML builds these images and
how you can customize them.
{% endhint %}

You can now run any ZenML pipeline using the Sagemaker orchestrator:
```shell
python file_that_runs_a_zenml_pipeline.py
```

### Run pipelines on a schedule

The ZenML Sagemaker orchestrator doesn't currently support running pipelines on
a schedule. We maintain a public roadmap for ZenML, which you can find
[here](https://zenml.io/roadmap). We welcome community contributions (see more
[here](https://github.com/zenml-io/zenml/blob/main/CONTRIBUTING.md)) so if you
want to enable scheduling for Sagemaker, please [do let us
know](https://zenml.io/slack-invite)!

### Additional configuration

For additional configuration of the Sagemaker orchestrator, you can pass
`SagemakerOrchestratorSettings` which allows you to configure (among others) the following attributes:

* `instance_type`: The instance type to use for the Sagemaker training job.
  (Defaults to `ml.t3.medium`.)

> * `execution_role: The IAM role to use for the Sagemaker training job.

Reviewer: ```suggestion
* `execution_role`: The IAM role to use for the Sagemaker training job.
```
